File: cafleet/src/cafleet/cli/message.py
"""``cafleet message`` — message broker commands."""

import click

from cafleet import broker, output
from cafleet.cli._helpers import (
    ensure_assets_current,
    fleet_id_option,
    from_member_id_option,
    full_flag,
    json_flag,
    member_id_option,
    quiet_flag,
    text_body_options,
    to_member_id_option,
)
from cafleet.cli._text_input import read_text_input


@click.group()
def message() -> None:
    """Message broker commands."""
    ensure_assets_current()


def _require_member_in_fleet(member_id: int, fleet_id: int) -> None:
    """Fleet-gate the acting member; runs before the command body."""
    if not broker.verify_member_fleet(member_id, fleet_id):
        raise click.ClickException(f"member {member_id} is not in fleet {fleet_id}.")


@message.command("send")
@fleet_id_option
@from_member_id_option
@to_member_id_option
@text_body_options("Message body (inline).")
@full_flag
@quiet_flag
@json_flag
@click.pass_context
def message_send(
    ctx, from_member_id, to_member_id, text, text_file, full, quiet, json_output
):
    """Send a unicast message to another member."""
    fleet_id = ctx.obj["fleet_id"]
    try:
        _require_member_in_fleet(from_member_id, fleet_id)
        body = read_text_input(text, text_file)
        result = broker.send_message(
            fleet_id,
            from_member_id,
            to_member_id,
            body,
        )
        output.truncate_message_text(result, full=full)
        rendered = output.render_messages_in_result(result, full=full)
        if json_output:
            click.echo(output.format_json(rendered))
        elif quiet:
            click.echo(str(result["message"]["message_id"]))
        else:
            click.echo("Message sent.\n" + output.format_message(result, full=full))
    except click.ClickException:
        raise
    except Exception as exc:
        raise click.ClickException(str(exc)) from exc


@message.command("broadcast")
@fleet_id_option
@from_member_id_option
@text_body_options("Message body (inline).")
@full_flag
@json_flag
@click.pass_context
def message_broadcast(ctx, from_member_id, text, text_file, full, json_output):
    """Broadcast a message to all members."""
    fleet_id = ctx.obj["fleet_id"]
    try:
        body = read_text_input(text, text_file)
        result = broker.broadcast_message(fleet_id, from_member_id, body)
        output.truncate_message_text(result, full=full)
        rendered = output.render_messages_in_result(result, full=full)
        if json_output:
            click.echo(output.format_json(rendered))
        elif full:
            click.echo(output.format_message(result[0]["message"], full=True))
        else:
            click.echo(
                f"broadcast id={result[0]['message']['message_id']} "
                f"recipients={result[0]['recipients']} "
                f"delivered={result[0]['delivered']}"
            )
    except click.ClickException:
        raise
    except Exception as exc:
        raise click.ClickException(str(exc)) from exc


@message.command("poll")
@fleet_id_option
@member_id_option
@full_flag
@json_flag
@click.pass_context
def message_poll(ctx, member_id, full, json_output):
    """Poll inbox for un-acked messages."""
    fleet_id = ctx.obj["fleet_id"]
    try:
        _require_member_in_fleet(member_id, fleet_id)
        result = broker.poll_messages(member_id)
        output.truncate_message_text(result, full=full)
        rendered = output.render_messages_in_result(result, full=full)
        if json_output:
            click.echo(output.format_json(rendered))
        else:
            click.echo(
                output.format_indexed_list(
                    result,
                    lambda t: output.format_message(t, full=full),
                    "No messages found.",
                )
            )
    except click.ClickException:
        raise
    except Exception as exc:
        raise click.ClickException(str(exc)) from exc


@message.command("ack")
@fleet_id_option
@member_id_option
@click.option("--message-id", type=int, required=True, help="Message ID to acknowledge")
@full_flag
@quiet_flag
@json_flag
@click.pass_context
def message_ack(ctx, member_id, message_id, full, quiet, json_output):
    """Acknowledge receipt of a message."""
    fleet_id = ctx.obj["fleet_id"]
    try:
        _require_member_in_fleet(member_id, fleet_id)
        result = broker.ack_message(member_id, message_id)
        output.truncate_message_text(result, full=full)
        rendered = output.render_messages_in_result(result, full=full)
        if json_output:
            click.echo(output.format_json(rendered))
        elif quiet:
            click.echo(str(result["message"]["message_id"]))
        else:
            click.echo(
                "Message acknowledged.\n" + output.format_message(result, full=full)
            )
    except click.ClickException:
        raise
    except Exception as exc:
        raise click.ClickException(str(exc)) from exc


@message.command("show")
@fleet_id_option
@member_id_option
@click.option("--message-id", type=int, required=True, help="Message ID to retrieve")
@full_flag
@json_flag
@click.pass_context
def message_show(ctx, member_id, message_id, full, json_output):
    """Get details of a specific message."""
    fleet_id = ctx.obj["fleet_id"]
    try:
        _require_member_in_fleet(member_id, fleet_id)
        result = broker.get_message(fleet_id, message_id)
        output.truncate_message_text(result, full=full)
        rendered = output.render_messages_in_result(result, full=full)
        if json_output:
            click.echo(output.format_json(rendered))
        else:
            click.echo(output.format_message(result, full=full))
    except click.ClickException:
        raise
    except Exception as exc:
        raise click.ClickException(str(exc)) from exc

File: cafleet/src/cli/message.rs
//! The `message` group and its shared handler sequence (SPEC §6.3 *message
//! group*): handler → emit fork (JSON = the complete untruncated result;
//! text = truncation + compact rendering). The broker derives the fleet and
//! recipient from the subject row.

use clap::{Args, Subcommand};
use serde_json::Value;

use super::helpers::{connect, resolve_body};
use crate::broker;
use crate::broker::records::NotificationAttempt;
use crate::config::Settings;
use crate::error::CafleetError;
use crate::output::{format_indexed_list, format_message, truncate_message_text};
use crate::presentation;
use crate::runtime::RuntimeNotifier;

#[derive(Args)]
#[group(required = true, multiple = false)]
pub(crate) struct BodyArgs {
    /// Inline message body. Exactly one of TEXT / --file.
    #[arg(value_name = "TEXT")]
    text: Option<String>,
    /// UTF-8 file carrying the body (`-` = stdin).
    #[arg(long, value_name = "PATH")]
    file: Option<String>,
}

#[derive(Subcommand)]
pub enum MessageCommand {
    /// Send a unicast message.
    Send {
        /// Sender's member ID.
        #[arg(long = "from-member-id")]
        from_member_id: i64,
        /// Recipient member ID.
        #[arg(long = "to-member-id")]
        to_member_id: i64,
        #[command(flatten)]
        body: BodyArgs,
        /// Output in JSON format.
        #[arg(long)]
        json: bool,
    },
    /// Broadcast a message to all fleet members.
    Broadcast {
        /// Sender's member ID.
        #[arg(long = "from-member-id")]
        from_member_id: i64,
        #[command(flatten)]
        body: BodyArgs,
        /// Output in JSON format.
        #[arg(long)]
        json: bool,
    },
    /// Fetch un-acked incoming messages.
    Poll {
        /// Recipient whose inbox is fetched.
        #[arg(value_name = "MEMBER_ID")]
        member_id: i64,
        /// Output in JSON format.
        #[arg(long)]
        json: bool,
    },
    /// Acknowledge a received message.
    Ack {
        /// Message to acknowledge.
        #[arg(value_name = "MESSAGE_ID")]
        message_id: i64,
        /// Output in JSON format.
        #[arg(long)]
        json: bool,
    },
    /// Show one message.
    Show {
        /// Message to fetch.
        #[arg(value_name = "MESSAGE_ID")]
        message_id: i64,
        /// Output in JSON format.
        #[arg(long)]
        json: bool,
    },
}

/// The emit fork of the shared handler sequence: JSON is the complete,
/// untruncated machine form; text truncates then renders compactly.
fn emit_result(
    settings: &Settings,
    mut result: Value,
    json: bool,
    text: impl FnOnce(&Value) -> String,
) {
    if json {
        println!("{}", crate::output::format_json(&result));
    } else {
        truncate_message_text(&mut result, settings.max_text_len);
        println!("{}", text(&result));
    }
}

pub fn run(settings: &Settings, command: MessageCommand) -> Result<(), CafleetError> {
    match command {
        MessageCommand::Send {
            from_member_id,
            to_member_id,
            body,
            json,
        } => {
            let text = resolve_body(body.text.as_deref(), body.file.as_deref(), "--file")?;
            let mut conn = connect(settings)?;
            let notifier = RuntimeNotifier::new(settings);
            let outcome = broker::send_message_record(
                &mut conn,
                &notifier,
                settings.max_text_len,
                from_member_id,
                &to_member_id.to_string(),
                &text,
            )?;
            if let NotificationAttempt::Failed { error: raw } = &outcome.notification {
                let message_id = outcome.message.message_id;
                return Err(CafleetError::App(format!(
                    "Message {message_id} was persisted, but pane notification failed: {raw}. \
                     Do not resend this message. Recover the recipient pane, then run \
                     'cafleet member ping {to_member_id}' or have the recipient run \
                     'cafleet message poll {to_member_id}'."
                )));
            }
            emit_result(
                settings,
                presentation::send_outcome(&outcome),
                json,
                |result| format!("Message sent.\n{}", format_message(result)),
            );
            Ok(())
        }
        MessageCommand::Broadcast {
            from_member_id,
            body,
            json,
        } => {
            let text = resolve_body(body.text.as_deref(), body.file.as_deref(), "--file")?;
            let mut conn = connect(settings)?;
            let notifier = RuntimeNotifier::new(settings);
            let result = broker::broadcast_message_record(
                &mut conn,
                &notifier,
                settings.max_text_len,
                from_member_id,
                &text,
            )?;
            emit_result(
                settings,
                Value::Array(vec![presentation::broadcast_outcome(&result)]),
                json,
                |_| {
                    format!(
                        "broadcast id={} recipients={} delivered={}",
                        result.message.message_id, result.recipients, result.delivered,
                    )
                },
            );
            Ok(())
        }
        MessageCommand::Poll { member_id, json } => {
            let conn = connect(settings)?;
            let result = broker::poll_message_records(&conn, member_id)?
                .iter()
                .map(presentation::message)
                .collect();
            emit_result(settings, Value::Array(result), json, |result| {
                let items = result.as_array().expect("poll returns a list");
                format_indexed_list(items, format_message, "No messages found.")
            });
            Ok(())
        }
        MessageCommand::Ack { message_id, json } => {
            let mut conn = connect(settings)?;
            let result =
                presentation::message_envelope(&broker::ack_message_record(&mut conn, message_id)?);
            emit_result(settings, result, json, |result| {
                format!("Message acknowledged.\n{}", format_message(result))
            });
            Ok(())
        }
        MessageCommand::Show { message_id, json } => {
            let conn = connect(settings)?;
            let result =
                presentation::message_envelope(&broker::get_message_record(&conn, message_id)?);
            emit_result(settings, result, json, format_message);
            Ok(())
        }
    }
}
